plenum/server/replica.py

```diff
@@ -1388,50 +1426,39 @@ def _apply_pre_prepare(self, pre_prepare: PrePrepare, sender: str) -> Optional[i
                 reqs.append(req)
             idx += 1
 
-        invalid_from_pp = invalid_index_serializer.deserialize(pre_prepare.discarded)
+        # 2. call callback for the applied batch
+        if self.isMaster:
+            three_pc_batch = ThreePcBatch.from_pre_prepare(pre_prepare,
+                                                           valid_txn_count=len(reqs) - len(invalid_indices),
+                                                           state_root=self.stateRootHash(pre_prepare.ledgerId,
+                                                                                         to_str=False),
+                                                           txn_root=self.txnRootHash(pre_prepare.ledgerId,
+                                                                                     to_str=False)
+                                                           )
+            self.node.onBatchCreated(three_pc_batch)
+
+        return reqs, invalid_indices, rejects
 
-        def revert():
-            self.revert(pre_prepare.ledgerId,
-                        old_state_root,
-                        len(pre_prepare.reqIdr) - len(invalid_from_pp))
+    def _validate_applied_pre_prepare(self, pre_prepare: PrePrepare,
+                                      reqs, invalid_indices, invalid_from_pp) -> Optional[int]:
         if len(invalid_indices) != len(invalid_from_pp):
-            if self.isMaster:
-                revert()
             return PP_APPLY_REJECT_WRONG
 
-        if pre_prepare.sub_seq_no != 0:
-            return PP_SUB_SEQ_NO_WRONG
-
-        if not pre_prepare.final:
-            return PP_NOT_FINAL
-
         digest = self.batchDigest(reqs)
-
-        # A PRE-PREPARE is sent that does not match request digest
         if digest != pre_prepare.digest:
-            if self.isMaster:
-                revert()
             return PP_APPLY_WRONG_DIGEST
```

Comment:
Just a small note - this check can be done before PREPREPARE is event applied. I remember we have some kind of static validation for messages (mostly schema checking) - can we move this check to that part somehow?

Reply:
Agree. Not in the scope of this work, but will be done in a separate PR.